src/include/memory/memory_reservation.hpp

```diff
@@ -16,79 +16,413 @@
+    const MemorySpace* getMemorySpace(const MemoryReservationManager& manager) const;
+    
+    //===----------------------------------------------------------------------===//
+    // Reservation Size Management
+    //===----------------------------------------------------------------------===//
+    
+    /**
+     * @brief Attempts to grow this reservation to a new larger size.
+     * @param new_size The new size for the reservation (must be larger than current size)
+     * @return true if the reservation was successfully grown, false otherwise
+     */
+    bool grow_to(size_t new_size);
+    
+    /**
+     * @brief Attempts to grow this reservation by additional bytes.
+     * @param additional_bytes Number of bytes to add to the current reservation
+     * @return true if the reservation was successfully grown, false otherwise
+     */
+    bool grow_by(size_t additional_bytes);
+    
+    /**
+     * @brief Attempts to shrink this reservation to a new smaller size.
+     * @param new_size The new size for the reservation (must be smaller than current size)
+     * @return true if the reservation was successfully shrunk, false otherwise
+     */
+    bool shrink_to(size_t new_size);
+    
+    /**
+     * @brief Attempts to shrink this reservation by removing bytes.
+     * @param bytes_to_remove Number of bytes to remove from the current reservation
+     * @return true if the reservation was successfully shrunk, false otherwise
+     */
+    bool shrink_by(size_t bytes_to_remove);
+    
+    // Disable copy/move to prevent issues with MemorySpace tracking
+    Reservation(const Reservation&) = delete;
+    Reservation& operator=(const Reservation&) = delete;
+    Reservation(Reservation&&) = delete;
+    Reservation& operator=(Reservation&&) = delete;
+    
+    ~Reservation() = default;
 };
 
+//===----------------------------------------------------------------------===//
+// MemoryReservationManager
+//===----------------------------------------------------------------------===//
+
+/**
+ * Central manager for memory reservations across multiple memory spaces.
+ * Implements singleton pattern and coordinates reservation requests using
+ * different strategies (specific space, tier-based, multi-tier fallback).
+ */
 class MemoryReservationManager {
 public:
-
-    static void initialize(const std::array<size_t, static_cast<size_t>(Tier::SIZE)>& tier_limits);
+    //===----------------------------------------------------------------------===//
+    // Configuration and Initialization
+    //===----------------------------------------------------------------------===//
+    
+    /**
+     * Configuration for a single MemorySpace.
+     * Contains all parameters needed to create a MemorySpace instance.
+     */
+    struct MemorySpaceConfig {
+        Tier tier;
+        size_t device_id;
+        size_t memory_limit;
+        std::vector<std::unique_ptr<rmm::mr::device_memory_resource>> allocators;
```

Comment:
When do we use the allocator in Memory Reservation Manager? I'm guessing we use the allocator in Memory Space to move data arounds?

Reply:
Just some stuff I have to clean up

Reply:
ok sorry I misread where this was. This is the allocators that are being set up for the MemorySpace.